File: 3-Archive/RSSReader-ObjC/RSSReader/RSSReaderGDiOSDelegate.m
/* Copyright (c) 2023 BlackBerry Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "RSSReaderGDiOSDelegate.h"
#import "Utilities.h"
#import "RSSManager.h"

@interface RSSReaderGDiOSDelegate()

@property(assign, nonatomic) BOOL hasAuthorized;

@end

@implementation RSSReaderGDiOSDelegate

-(void)setrssReaderAppDelegate:(AppDelegate *)rssReaderAppDelegate
{
    _rssReaderAppDelegate = rssReaderAppDelegate;
    [self didAuthorize];
}

-(void)didAuthorize
{
    if (self.hasAuthorized && self.rssReaderAppDelegate) {
        NSLog(@"%s", __FUNCTION__);
    }
    
    GDAuthDelegateInfo* gdAuthDelegateInfo = [[GDiOS sharedInstance] getAuthDelegate];

    NSLog(@"didAuthorize getAuthDelegate Name = %@, address = %@, applicationId = %@, isAuthenticationDelegated = %@",
          gdAuthDelegateInfo.name, gdAuthDelegateInfo.address, gdAuthDelegateInfo.applicationId,
          gdAuthDelegateInfo.isAuthenticationDelegated ? @"YES" : @"NO");
}

+(instancetype)sharedInstance
{
    static RSSReaderGDiOSDelegate *rssReaderGDiOSDelegate = nil;
    static dispatch_once_t onceToken = 0;
    dispatch_once(&onceToken, ^{
        rssReaderGDiOSDelegate = [[RSSReaderGDiOSDelegate alloc] init];
    });
    return rssReaderGDiOSDelegate;
}

#pragma mark - Good Dynamics Delegate Methods

-(void)handleEvent:(GDAppEvent*)anEvent
{
    /* Called from _good when events occur, such as system startup. */
    
    switch (anEvent.type)
    {
        case GDAppEventAuthorized:
        {
            [self onAuthorized:anEvent];
            break;
        }
        case GDAppEventNotAuthorized:
        {
            [self onNotAuthorized:anEvent];
            break;
        }
        case GDAppEventRemoteSettingsUpdate:
        {
            // handle app config changes
            break;
        }
        case GDAppEventServicesUpdate:
        {
            //A change to services-related configuration.
            break;
        }
        case GDAppEventPolicyUpdate:
        {
            //A change to one or more application-specific policy settings has been received.
            break;
        }
        case GDAppEventEntitlementsUpdate:
        {
            //A change to the entitlements data has been received.
            break;
        }
        default:
        {
            NSLog(@"event not handled: %@", anEvent.message);
            break;
        }
    }
}

-(void) onNotAuthorized:(GDAppEvent*)anEvent
{
    /* Handle the Good Libraries not authorized event. */
    
    switch (anEvent.code) {
        case GDErrorActivationFailed:
        case GDErrorProvisioningFailed:
        case GDErrorPushConnectionTimeout:
        case GDErrorSecurityError:
        case GDErrorAppDenied:
        case GDErrorAppVersionNotEntitled:
        case GDErrorBlocked:
        case GDErrorWiped:
        case GDErrorRemoteLockout:
        case GDErrorPasswordChangeRequired: {
            // an condition has occured denying authorization, an application may wish to log these events
            NSLog(@"onNotAuthorized %@", anEvent.message);
            break;
        }
        case GDErrorIdleLockout: {
            // idle lockout is benign & informational
            break;
        }
        default:
            NSAssert(false, @"Unhandled not authorized event");
            break;
    }
}

-(void) onAuthorized:(GDAppEvent*)anEvent
{
    /* Handle the Good Libraries authorized event. */
    
    switch (anEvent.code) {
        case GDErrorNone: {
            if (!self.hasAuthorized) {
                // launch application UI here
                UIWindow *appWindow = self.rssReaderAppDelegate.window;
                appWindow.tintColor = RGB(52.f, 69.f, 84.f);
                //Detect the device
                if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone)
                {
                    /*
                     *	iPhone start - based on single UINavigationController
                     */
                    // Setup the rootviewcontroller and a navigation controller (for the splitviewcontroller)
                    self.navController = [[Utilities storyboard] instantiateViewControllerWithIdentifier:@"iPhoneNC"];
                    
                    appWindow.rootViewController = _navController;
                }
                else
                {
                    
                    /*
                     *	iPad start - based on UISplitViewController
                     */
                    self.splitViewController = [[Utilities storyboard] instantiateViewControllerWithIdentifier:@"SplitVC"];
                    
                    self.navController = [self.splitViewController.viewControllers firstObject];
                    self.detailNavigationController = [self.splitViewController.viewControllers lastObject];
                    self.rssReaderAppDelegate.window.rootViewController = self.splitViewController;
                    
                }
                [[RSSManager sharedRSSManager] setAlertPresenter:self.navController];
                self.hasAuthorized = YES;
                [self didAuthorize];
            }
            break;
        }
        default:
            NSAssert(false, @"Authorized startup with an error");
            break;
    }
}


@end

File: 3-Archive/RSSReader/RSSReaderSwift/RSSReaderGDiOSDelegate.swift
/* Copyright (c) 2023 BlackBerry Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import Foundation
import UIKit
import BlackBerryDynamics.Runtime


class RSSReaderGDiOSDelegate: NSObject, GDiOSDelegate {
    
    static let sharedInstance = RSSReaderGDiOSDelegate()
    weak var appDelegate: AppDelegate? {
        didSet {
            didAuthorize()
        }
    }
    weak var rootViewController: UIViewController?{
        didSet {
            didAuthorize()
        }
    }
    
    var hasAuthorized:Bool = false
    
    
    fileprivate override init() {
        print(#file, #function)
        super.init()
    }
    
    deinit {
        print(#file, #function)
    }
    
    
    func didAuthorize() -> Void {
        if (hasAuthorized && (rootViewController != nil) && (appDelegate != nil) ) {
            self.appDelegate!.didAuthorize()
        }
    }
    
    func handle(_ anEvent:GDAppEvent) {
        
        switch anEvent.type
        {
        case .authorized:
            self.onAuthorized(anEvent);
            break;
            
        case .notAuthorized:
            self.onNotAuthorized(anEvent);
            break;
            
        case .remoteSettingsUpdate:
            //A change to application-related configuration or policy settings.
            break;
            
        case .servicesUpdate:
            //A change to services-related configuration.
            break;
            
        case .policyUpdate:
            //A change to one or more application-specific policy settings has been received.
            break;
            
        default:
            print("Event not handled: '\(anEvent.message)'")
            break
        }
    }
    
    func onNotAuthorized(_ anEvent:GDAppEvent ) {
        /* Handle the Good Libraries not authorized event. */
        switch anEvent.code {
        case .errorActivationFailed: break
        case .errorProvisioningFailed: break
        case .errorPushConnectionTimeout: break
        case .errorSecurityError: break
        case .errorAppDenied: break
        case .errorAppVersionNotEntitled: break
        case .errorBlocked: break
        case .errorWiped: break
        case .errorRemoteLockout: break
        case .errorPasswordChangeRequired:
            // an condition has occured denying authorization, an application may wish to log these events
            print("onNotAuthorized \(anEvent.message)")
        case .errorIdleLockout:
            // idle lockout is benign & informational
            break
            
        default: assert(false, "Unhandled not authorized event");
        }
    }
    
    func onAuthorized(_ anEvent: GDAppEvent) {
        // Handle the Good Libraries not authorized event
        switch anEvent.code {
        case .errorNone :
            if (!hasAuthorized) {
                // Prepare main storyboard and root view controller
                let mainStoryboard: UIStoryboard = UIStoryboard(name: "Main", bundle: Bundle.main)
                let rootViewController = mainStoryboard.instantiateInitialViewController()
                self.rootViewController = rootViewController;

                self.appDelegate!.window?.rootViewController = rootViewController
                hasAuthorized = true
                didAuthorize()
            }
        default:
            assert(false, "Authorized startup with an error")
            break
        }
    }
}
